service-base/src/main/java/org/eclipse/hono/service/auth/device/SubjectDnCredentials.java

```diff
@@ -43,6 +43,7 @@ private SubjectDnCredentials(final String tenantId, final String authId, final J
      * @param subjectDn The subject DN of the device's client certificate.
      * @return The credentials.
      * @throws NullPointerException if any of the parameters are {@code null}.
+     * @deprecated Use {@link #create(String, String, JsonObject)} instead.
```

Comment:
Yes, an empty client context is still valid. Since I have added a new version with a client context parameter for each *create* method, I think the old methods should be removed: a) the new methods are more explicit, b) the old ones are no longer used by Hono, c) if you do use them, it is harder to tell if it is intended not to provide the client context (in this case the AMQP adapter actually used one of them and therefore inadvertently does not supported auto-provisioning).

My general thoughts on this topic: It is not yet decided which part of the code should be seen as API. By adding new versions of the methods that expose the client context, the old versions have become convenient methods. I prefer not to have many convenient methods in an API, as they may not age well and add unnecessary complexity. If we just add new code, we will be faced with a massive increase in complexity without, as in this case, an increased set of features.
If it is not part of the API, the unused methods could be removed immediately.

Reply:
I see your point but in this case I also do not see any risk of overloading the class with convenience factory methods. On the contrary, I do not see why I should need to pass an empty JsonObject if there is no client context ...